rust_src/src/frames.rs

```diff
@@ -614,17 +628,50 @@ pub fn frame_focus(frame: LispFrameLiveOrSelected) -> LispObject {
 /// means do not activate FRAME.
 ///
 /// If there is no window system support, this function does nothing.
-#[lisp_fn(min = "1")]
-pub fn x_focus_frame(_frame: LispFrameLiveOrSelected, _noactivate: bool) -> LispObject {
+#[lisp_fn(min = "1", name = "x-focus-frame", c_name = "x_focus_frame")]
+pub fn x_focus_frame_lisp(_frame: LispFrameLiveOrSelected, _noactivate: bool) {
     #[cfg(feature = "window-system")]
     {
         let mut frame_ref: LispFrameRef = _frame.into();
         unsafe {
-            c_x_focus_frame(frame_ref.as_mut(), _noactivate);
+            x_focus_frame(frame_ref.as_mut(), _noactivate);
         }
     }
+}
 
-    Qnil
+/// Return a list of all live frames.
+/// The return value does not include any tooltip frame.
+#[lisp_fn]
+pub fn frame_list() -> LispObject {
+    #[cfg(feature = "window-system")]
+    {
+        let mut list: LispObject = Qnil;
+        for_each_frame!(f => {
+            if !f.has_tooltip() {
+                list = (f, list).into();
+            }
+        }
+        );
+        // Reverse list for consistency with the !HAVE_WINDOW_SYSTEM case.
+        unsafe { Fnreverse(list) }
+    }
+    #[cfg(not(feature = "window-system"))]
+    {
+        unsafe { Fcopy_sequence(Vframe_list) }
+    }
+}
+
+/// Return a list of all frames now \"visible\" (being updated).
+#[lisp_fn]
+pub fn visible_frame_list() -> LispObject {
+    let mut list: LispObject = Qnil;
+    for_each_frame!(f => {
+            if f.is_visible() {
+                    list = (f, list).into();
+            }
+        }
+        );
+    list
```

Comment:
`visible_frame_list` and `frame_list` are the same code with the only difference being the predicate used to decide which elements are included in the list. Perhaps the right thing to do is factor that code out.

```
fn filter_frame_list(predicate: impl Fn(LispFrameRef) -> bool) -> LispObject {
    // or the fold version...
    for_each_frame!(f => {
            if predicate(f) {
                list = (f, list).into();
            }
        });
}

pub fn visible_frame_list() -> LispObject {
    filter_frame_list(|f| f.is_visible())
}

pub fn frame_list() -> LispObject {
    #[cfg(feature = "window-system")]
    {
        let mut list = filter_frame_list(|f| !f.has_tooltip());
        // Reverse list for consistency with the !HAVE_WINDOW_SYSTEM case.
        unsafe { Fnreverse(list) }
    }
    #[cfg(not(feature = "window-system"))]
    {
        unsafe { Fcopy_sequence(Vframe_list) }
    }
}
```

Reply:
I think we should use `for_each_frame` consistently or change it to something else. So I will take this one.

Reply:
You asked about folds recently. So I pointed these out. The for each is fine. And sometimes it reads better. These two versions are nearly identical.